flink/v1.18/flink/src/main/java/org/apache/iceberg/flink/FlinkFileIO.java

```diff
@@ -0,0 +1,182 @@
+    return new FlinkOutputFile(new Path(path));
+  }
+
+  @Override
+  public void deleteFile(String path) {
+    Path toDelete = new Path(path);
+    try {
+      toDelete.getFileSystem().delete(toDelete, false /* not recursive */);
+    } catch (IOException e) {
+      throw new UncheckedIOException(String.format("Failed to delete file: %s", path), e);
+    }
+  }
+
+  @Override
+  public Iterable<FileInfo> listPrefix(String prefix) {
+    LOG.debug("Listing {}", prefix);
+    Path prefixToList = new Path(prefix);
+    try {
+      return listPrefix(prefixToList.getFileSystem(), prefixToList);
+    } catch (IOException e) {
+      throw new UncheckedIOException(e);
+    }
+  }
+
+  @Override
+  public void deletePrefix(String prefix) {
+    Path prefixToDelete = new Path(prefix);
+
+    try {
+      prefixToDelete.getFileSystem().delete(prefixToDelete, true /* recursive */);
+    } catch (IOException e) {
+      throw new UncheckedIOException(e);
+    }
+  }
+
+  @Override
+  public void deleteFiles(Iterable<String> pathsToDelete) throws BulkDeletionFailureException {
+    AtomicInteger failureCount = new AtomicInteger(0);
+    Tasks.foreach(pathsToDelete)
+        .executeWith(executorService())
+        .retry(DELETE_RETRY_ATTEMPTS)
+        .stopRetryOn(FileNotFoundException.class)
+        .suppressFailureWhenFinished()
+        .onFailure(
+            (f, e) -> {
+              LOG.error("Failure during bulk delete on file: {} ", f, e);
+              failureCount.incrementAndGet();
+            })
+        .run(this::deleteFile);
+
+    if (failureCount.get() != 0) {
+      throw new BulkDeletionFailureException(failureCount.get());
+    }
+  }
+
+  @Override
+  public void initialize(Map<String, String> props) {
+    this.properties = SerializableMap.copyOf(props);
+  }
+
+  @Override
+  public Map<String, String> properties() {
+    return properties.immutableMap();
+  }
+
+  private Iterable<FileInfo> listPrefix(FileSystem fs, Path fileName) {
+    try {
+      FileStatus[] statuses = fs.listStatus(fileName);
+      LOG.debug("Listing path {} {}", fileName, fs.listStatus(fileName));
+      if (statuses == null) {
+        // Check the file is there and ready. If so, then we can assume this is an empty dir.
+        fs.getFileStatus(fileName);
+        statuses = new FileStatus[0];
+      }
+
+      return Iterables.concat(
+          Arrays.stream(statuses)
+              .map(
+                  fileStatus -> {
+                    if (fileStatus.isDir()) {
+                      return listPrefix(fs, fileStatus.getPath());
+                    } else {
+                      return Collections.singleton(
+                          new FileInfo(
+                              fileStatus.getPath().toString(),
+                              fileStatus.getLen(),
+                              fileStatus.getModificationTime()));
+                    }
+                  })
+              .collect(Collectors.toSet()));
```

Comment:
do we really need to `Collectors.toSet()` ? Are we attempting to deduplicate items with this statement ?

Reply:
Changed to `toList`